command/agent/snapshot_endpoint.go

```diff
@@ -0,0 +1,45 @@
+package agent
+
+import (
+	"net/http"
+
+	"github.com/hashicorp/consul/consul/structs"
+)
+
+// Snapshot handles requests to take and restore snapshots. This uses a special
+// mechanism to make the RPC since we potentially stream large amounts of data
+// as part of these requests.
+func (s *HTTPServer) Snapshot(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	var args structs.SnapshotRequest
+	s.parseDC(req, &args.Datacenter)
```

Comment:
I didn't add the full set of `QueryOptions` in here since not all of them are supported.